### DISCUSSION TITLE
table checkbox and bulk update

### DISCUSSION THREAD

**Discussion by samg666** [Budibase/budibase#3034] [Help]
Hello,

I am trying to add a checkbox on the table, then bulk update specific field. It seems there's no something like checkedRows/selectedRows property to bind, and there no On Click action on the checkbox. I don't know if there any way can do this request?

![iShot2021-10-18 13 55 16](https://user-images.githubusercontent.com/29649651/137676343-baf2516d-e700-421a-919e-d907acda68db.png)

**Comment by mjashanks** [comment]
Hey @ggs331. Unfortunately there is no way to do this right now. 

"Selected rows" been raised before:
#1723
#1598

So, we'd like to get it done.

As for Bulk update - not possible with Budibase's internal DB at the moment - only with a SQL custom query.

**Reply by samg666** [reply]
Hello @mjashanks , actually I want have the bulk update on external MySQL.  I am thinking if there's a way to get selected rows's id field, then I may be able to use custom SQL with selected id as parameters, and update via something like `UPDATE tickets SET department='Sales' WHERE id IN {{selectedIDs}}`. But the problem is, I didn't find a way to get selected id. Can you share some idea how to do a bulk update with SQL custom query?

**Reply by mjashanks** [reply]
Yea - I don't think I have anything more to share than you have just mentioned!

All I was going to say was `UPDATE table SET field WHERE ....`, but there's no way to pass down a list of `id`s. The `where` clause would have to be some other criteria.

**Reply by samg666** [reply]
Yes, I didn't find a way to pass the selected `ids`. It will be possible if the checkbox has the on click action like the button, then I will be able to store the Table.DataProvider.id to App State, then pass it as custom SQL parameter.

**Reply by mjashanks** [reply]
Yea, another way we could enable this is by allowing the `key` of appstate to be bindable
![image](https://user-images.githubusercontent.com/3524181/137712965-7f517927-a62f-4c10-b766-3ad463be36a7.png)

meaning you could see a `key` of `isSelected_{{ Row._id }}`. Not possible right now though.

**Reply by samg666** [reply]
hope to have this kind of feature soon

**Comment by marco-santulli** [comment]
hi @mjashanks 

I wanted to take this discussion to ask something.  I have not found any way to get the list of "selected id's" in the table from a binding.  Also, even if I did, how would one pass it to the query underneath? 

I think a workaround that would help resolve 80% of this class of problems would be to be able to:
1. bind to the list of selected ids from a button action or elsewhere (so should be visible at the page level)
2. make queries callable as a binding from javascript.  

So one would be able to do somethinking like this by "executing a javascript function" on a button action:
a. get list of selected ids from the table
b. iterate over the list calling "queryX with the ID parameter".

what do you think?

**Reply by mjashanks** [reply]
Hey Marco. 

I actually just answered a question about this - I was able to use selected rows, just as you mentioned: https://github.com/Budibase/budibase/discussions/4897#discussioncomment-2358124

The thing to note about it at the moment is that 
- When `Selected rows` is used in handlebars, it converts to a comma separated string.
- When used in Javascript, it returns the array of IDs

Almost always, I'd suggest using javascript.

If you want to rehydrate the IDs into data again, you could use the structure
```
> Data Provider
   > Table (selected rows)
   > Button 
```

The button action would take the selected rows and query against `Data Provider.Rows` to get the full data again. Then we could use that data in subsequent actions.

**Reply by marco-santulli** [reply]
Hi Mike,

thanks, I have in fact been able to get to the selected rows.  I think a small example of how to deal with the data would be useful, and in particular, if it is possible to access a stadard query (like a select) from javascript like we do for the selected rows, in that way I could re-hydrate without going to the front-end?

M

**Reply by Havock94** [reply]
> The button action would take the selected rows and query against `Data Provider.Rows` to get the full data again. Then we could use that data in subsequent actions.

Hello, I would like to ask a question about this. I'm trying to create a screen where a button deletes the selected rows in a table.

While I managed to get a list (array) of `_id`s and `_rev`s from the selected rows, I know don't know how to use them to trigger multiple rows deletion.
As far as I can see, both row delete action and automation only accept one `id` and one `rev` at a time, and there's no way to iterate array to repeat the same delete action for each element of the array, am I right?

Thanks!

**Reply by mjashanks** [reply]
Hi @Havock94 - I believe that this should be possible by the end of April, due to this: #4551 and #4427 and #5145

It will be done by sending a list of IDs to an automation, then running an automation `For Each` block.

**Reply by aderici** [reply]
sharing a code example to accomplish what @mjashanks mentioned, in your button you can use a logic and write a similar code. This example takes selected rows, queries data source to retrieve Auto ID column values, redirects user to a new url with auto ids comma separated as a url parameter : 

```
let myRows = $("Samples Table.Selected rows")
let myData = $("Data Provider.Rows")
let mySelectedRows = myData.filter((row)=>{return myRows.includes(row._id)})
return "/samples/confirm/" + mySelectedRows.map((row)=>{return row["Auto ID"]}).join(",")
```

**Comment by ishani-tagore** [comment]
I am trying to feed the outputs of @aderici 's code into the SQL query below as a replacement for "VALUES (1), (2), (3)":
SELECT {{Project_id}}, label_id
FROM **(VALUES (1), (2), (3))** AS label_ids(label_id);

The format of @aderici 's output is an array, so I modified the code to work with the SQL above, but still get a syntax error when running query:
let myRows = $("Master List of Labels.Selected rows");
let myData = $("New Data Provider.Rows");
let mySelectedRows = myData.filter((row) => myRows.includes(row._id));
let resultArray = mySelectedRows.map((row) => row["id"]);
let formattedList = resultArray.map((value) => (${value})).join(", ");
return formattedList;

The results of the javascript above are displayed in (2), (3), (7), but produce a syntax error in the query
<img width="741" alt="image" src="https://github.com/Budibase/budibase/assets/94510833/7c95a218-cf1c-4b23-932e-e25aa6f86ae7">

**Reply by aderici** [reply]
MYSQL IN operator uses a syntax like : value IN (value1, value2, value3,...) so maybe something like 

let myInClause = "'" + resultArray.join("','") + "'"

... WHERE XYZ IN {{myInClause}}

**Reply by ishani-tagore** [reply]
Tried that; didn't work :( I tried the following three formats of the SQL input clause and for all, I got "query syntax error at $2": 

**Format: '5', '3', '7'**
let mySelectedRows = myData.filter((row) => myRows.includes(row._id));
let resultArray = mySelectedRows.map((row) => row["id"]);
let myInClause = "'" + resultArray.join("','") + "'";
return myInClause;

**Format: (5, 3, 7)**
let mySelectedRows = myData.filter((row) => myRows.includes(row._id));
let resultArray = mySelectedRows.map((row)=>{return row["id"]}).join(", ")
return resultArray;

**Format: (5), (3), (7)**
let resultArray = mySelectedRows.map((row) => row["id"]);
let formattedList = resultArray.map((value) => `(${value})`).join(", ");
return formattedList;
